gateway/handler.go

```diff
@@ -887,6 +889,103 @@ func (i *handler) handleSuperfluousNamespace(w http.ResponseWriter, r *http.Requ
 	return true
 }
 
+// Detect 'Ipfs-Path-Affinity' (IPIP-462) headers in request and use values as a content
+// routing hints if passed paths are not already in the local datastore.
+// These optional hints are mostly useful for trustless block requests.
+// See https://github.com/ipfs/specs/pull/462
+func (i *handler) handlePathAffinityHints(w http.ResponseWriter, r *http.Request, contentPath path.Path, logger *zap.SugaredLogger) {
+	headerName := "Ipfs-Path-Affinity"
+	// Skip if no header
+	if r.Header.Get(headerName) == "" {
+		return
+	}
+	// Skip if contentPath is already locally cached
+	if i.backend.IsCached(r.Context(), contentPath) {
+		return
+	}
+	// Check canonical header name
+	// NOTE: we don't use r.Header.Get() because client can send this header more than once
+	headerValues := r.Header[headerName]
+	// If not found, try lowercase version.
+	// NOTE: this is done manually because direct key access does not come with canonicalization, like Header.Get() does
+	if len(headerValues) == 0 {
+		headerValues = r.Header[strings.ToLower(headerName)]
+	}
+
+	// Limit the headerValues to the first 3 items (abuse protection)
+	if len(headerValues) > 3 {
+		headerValues = headerValues[:3]
+	}
+
+	// Process affinity hints
+	for _, headerValue := range headerValues {
+		// Non-ascii paths are percent-encoded.
+		// Decode if the value starts with %2F (percent-encoded '/')
+		if strings.HasPrefix(headerValue, "%2F") {
+			decodedValue, err := url.PathUnescape(headerValue)
+			if err != nil {
+				logger.Debugw("skipping invalid Ipfs-Path-Affinity hint", "error", err)
+				continue
+			}
+			headerValue = decodedValue
+		}
+		// Confirm it is a valid content path
+		affinityPath, err := path.NewPath(headerValue)
+		if err != nil {
+			logger.Debugw("skipping invalid Ipfs-Path-Affinity hint", "error", err)
+			continue
+		}
+
+		// Skip duplicated work if immutable affinity hint is a subset of requested immutable contentPath
+		// (protect against broken clients that use affinity incorrectly)
+		if !contentPath.Mutable() && !affinityPath.Mutable() && strings.HasPrefix(contentPath.String(), affinityPath.String()) {
```

Comment:
The comment is a little confusing about which if a prefix of the other.
> if immutable affinity hint is a subset of requested immutable contentPath

It could mean that  the contentPath should be the prefix, since the subset is a more specific (longer) path:
```
contentPath = "/top-level"
affinityPath = "/top-level/sub-path"
```
or it could me that the affinity path is a _string subset_ of the longer content path:
```
contentPath = "/top-level/stuff/sub-stuff"
affinityPath = "/top-level/stuff"
```